### PR TITLE
Use primary terms as authority to fail shards

### PR DESCRIPTION
**Pull request by ywelsch**
A primary shard currently instructs the master to fail a replica shard that it fails to replicate writes to before acknowledging the writes to the client. To ensure that the primary instructing the master to fail the replica is still the current primary in the cluster state on the master, it submits not only the identity of the replica shard to fail to the master but also its own shard identity. This can be problematic however when the primary is relocating. After primary relocation handoff but before the primary relocation target is activated, the primary relocation target is replicating writes through the authority of the primary relocation source. This means that the primary relocation target should probably send the identity of the primary relocation source as authority. However, this is not good enough either, as primary shard activation and shard failure instructions can arrive out-of-order. This means that the relocation target would have to send both relocation source and target identity as authority. Fortunately, there is another concept in the cluster state that represents this joint authority, namely primary terms. The primary term is only increased on initial assignment or when a replica is promoted. It stays the same however when a primary relocates.

This PR changes ShardStateAction to rely on primary terms for shard authority. It also changes the wire format to only transmit `ShardId` and allocation id of the shard to fail (instead of the full `ShardRouting`), so that the same action can be used in a subsequent PR to remove allocation ids from the active allocation set for which there exist no `ShardRouting` in the cluster anymore. Last but not least, this PR also makes AllocationService less lenient, requiring ShardRouting instances that are passed to its applyStartedShards and applyFailedShards methods to exist in the routing table. ShardStateAction, which is calling these methods, now has the responsibility to resolve the ShardRouting objects that are to be started / failed, and remove duplicates.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 1]
```diff
@@ -308,6 +308,11 @@ public static boolean isConflictException(Throwable t) {
         ShardRouting routingEntry();
 
         /**
+         * primary term for this primary (fixed when primary shard is created)
+         */
+        long primaryTerm();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: this also means that we can fold setting the primary term on a request into the ReplicationOperation - we currently only assert that it was set.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 2]
```diff
@@ -234,51 +232,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                            task.shardId,
+                            "primary term [" + task.primaryTerm + "] did not match current primary term [" +
+                                indexMetaData.primaryTerm(task.shardId.id()) + "]"));
+                        continue;
+                    }
 
-            // tasks that correspond to non-existent shards are marked
-            // as successful
-            batchResultBuilder.successes(partition.getOrDefault(ValidationResult.SHARD_MISSING, Collections.emptyList()));
+                    ShardRouting matched = currentState.getRoutingNodes().getByAllocationId(task.shardId, task.allocationId);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: why is this going through RoutingNodes ? isn't routing table more natural (and doesn't require building the routing nodes variant)?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 3]
```diff
@@ -234,51 +232,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
 
-            // partition tasks into those that correspond to shards
-            // that exist versus do not exist
-            Map<ValidationResult, List<ShardRoutingEntry>> partition =
-                tasks.stream().collect(Collectors.groupingBy(task -> validateTask(currentState, task)));
+            for (ShardRoutingEntry task : tasks) {
+                IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
+                if (indexMetaData == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
+                    batchResultBuilder.success(task);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we add some trace logging here? I can imagine it will save some WTF at some point.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 4]
```diff
@@ -234,51 +232,50 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                        continue;
+                    }
 
-            // tasks that correspond to non-existent shards are marked
-            // as successful
-            batchResultBuilder.successes(partition.getOrDefault(ValidationResult.SHARD_MISSING, Collections.emptyList()));
+                    ShardRouting matched = currentState.getRoutingNodes().getByAllocationId(task.shardId, task.allocationId);
+                    if (matched == null) {
+                        // tasks that correspond to non-existent shards are marked as successful
+                        batchResultBuilder.success(task);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we add some trace logging here too?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 5]
```diff
@@ -403,48 +366,54 @@ public void onFailure(String source, Exception e) {
...
+            this.primaryTerm = primaryTerm;
             this.message = message;
             this.failure = failure;
         }
 
-        public ShardRouting getShardRouting() {
-            return shardRouting;
+        public boolean matches(ShardRouting shardRouting) {
+            return shardId.equals(shardRouting.shardId()) && shardRouting.assignedToNode() &&
+                allocationId.equals(shardRouting.allocationId().getId());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can the aid matching the implementation and the rest just assertions ? it should be enough

**Comment by bleskes** [comment]
Thanks @ywelsch . I started having a look but I wonder why does the move from a complete shard routing for the failing/starting shard need to be remove in favor of the new ShardAllocationId class. I like the full ShardRouting because it allows us to log more debug/throw exception with more information when needed. What does the extra class buy us?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 6]
```diff
@@ -131,15 +128,16 @@ private static boolean isMasterChannelException(TransportException exp) {
     /**
      * Send a shard failed request to the master node to update the
      * cluster state.
-     *  @param shardRouting       the shard to fail
-     * @param sourceShardRouting the source shard requesting the failure (must be the shard itself, or the primary shard)
+     * @param shardRouting       the shard to fail
+     * @param primaryTerm        the primary term associated with the primary shard when the shard is failed by the primary. Use 0L if the
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: Using 0 as a non-value (because we serialize using out.writeVLong ) seems brittle to me. I rather use -1 and serialize using normal `out.writeLong`. With 0 we will always be asking ourselves if it's a valid value.  We can also potentially have two methods here - failReplica and failShard, where failReplica requires a valid primary term and failShard doesn't take one (but uses -1 internally) so users won't need to know about this implementation detail.

**Comment by ywelsch** [comment]
@bleskes In a follow-up PR, I want to extend `ShardStateAction` with the capability for primary shards to mark shard copies as stale, some of which don't have a ShardRouting, but for which we only know ShardId and allocation id. The suggested interface here gives this uniform handling, and allows batching with shard failed as well. Furthermore, the `ShardAllocationId` provides a lean interface for AllocationService.applyStartedShards / applyFailedShards, making it clear what information is actually used, and allows code sharing between the two (e.g. to resolve shard) as well as the to be introduced method to remove active allocation ids. Regarding additional debug information, logging ShardRouting at sender site and task information to correlate requests to receiver should be enough I think?

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: conversely, we could as well use the primaryTerm on the replicaRequest when failing the shard. This means that we don't need to extend the `Primary` interface with `primaryTerm`.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: +1 to explore that option. Although, I always doubted whether we should set the primary terms on the ReplicationOperation level, I didn't do it because it wasn't needed and reluctantly opted to keep things as simple as possible and left it at that assert we have now. With the move the primary term based failures we bring the terms into scope, which makes me more inclined to also set them. As said, I'm still on the fence about it myself so I will go along with not doing it, if you feel differently.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 7]
```diff
@@ -225,22 +224,30 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
...
+        // add nodes of failing shards (that exist) to ignore list
+        for (FailedRerouteAllocation.FailedShard failedShard : failedShards) {
+            ShardRouting shardRouting = resolveShardRouting(allocation, "shard failure", failedShard.shardId(), failedShard.allocationId());
+            if (shardRouting != null) {
+                allocation.addIgnoreShardForNode(shardRouting.shardId(), shardRouting.currentNodeId());
+            }
+        }
+
         boolean changed = false;
-        // as failing primaries also fail associated replicas, we fail replicas first here so that their nodes are added to ignore list
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: why do you feel this is not important?

**Comment by bleskes** [comment]
@ywelsch discussed this and I asked to remove `ShardAllocationId` in favor of doing  all resolving in `ShardStateAction`. This also means that `AllocationService` will become strict and can expect the callers to supply shard routings that actually exist rather than being lenient try to resolve what it can.

**Comment by ywelsch** [comment]
@bleskes I've pushed 3c0172b that makes `AllocationService` strict, expecting callers to supply shard routings that exist in the routing table.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 8]
```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+            for (ShardEntry task : tasks) {
+                IndexMetaData indexMetaData = currentState.metaData().index(task.shardId.getIndex());
+                if (indexMetaData == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
+                    logger.trace("{} index metadata for index {} does not exist anymore, marking task {} as successful", task.shardId,
+                        task.shardId.getIndex(), task);
+                    batchResultBuilder.success(task);
+                } else {
+                    // non-local requests
+                    if (task.primaryTerm > 0 && task.primaryTerm != indexMetaData.primaryTerm(task.shardId.id())) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we assert in the if clause that the primary term is smaller than the current one?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 9]
```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                        logger.trace("{} primary term does not match for task {} (required [{}] but was [{}])", task.shardId,
+                            indexMetaData.primaryTerm(task.shardId.id()), task.primaryTerm, task);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: nit: put `indexMetaData.primaryTerm(task.shardId.id())` in a variable?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 10]
```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                    ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                    if (matched == null) {
+                        // tasks that correspond to non-existent shards are marked as successful
+                        logger.trace("{} shard with allocation id {} does not exist anymore, marking task {} as successful", task.shardId,
+                            task.allocationId, task);
+                        batchResultBuilder.success(task);
+                    } else {
+                        // remove duplicate actions
+                        if (seenValidShardRoutings.contains(matched)) {
+                            logger.trace("{} shard with allocation id {} exists and already scheduled, processing task {}", task.shardId,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: :) maybe add to the comment that AllocationService expects a clean list?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 11]
```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
-        public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
-            BatchResult.Builder<ShardRoutingEntry> batchResultBuilder = BatchResult.builder();
-
-            // partition tasks into those that correspond to shards
-            // that exist versus do not exist
-            Map<ValidationResult, List<ShardRoutingEntry>> partition =
-                tasks.stream().collect(Collectors.groupingBy(task -> validateTask(currentState, task)));
+        public BatchResult<ShardEntry> execute(ClusterState currentState, List<ShardEntry> tasks) throws Exception {
+            BatchResult.Builder<ShardEntry> batchResultBuilder = BatchResult.builder();
+            List<ShardEntry> validTasks = new ArrayList<>();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: maybe call this pendingTasks or resolvedTasks? I got a bit confused by the valid notion - some of the tasks are marked as successful but are not valid :)

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 12]
```diff
@@ -370,27 +368,54 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
...
-        @Override
-        public BatchResult<ShardRoutingEntry> execute(ClusterState currentState, List<ShardRoutingEntry> tasks) throws Exception {
-            BatchResult.Builder<ShardRoutingEntry> builder = BatchResult.builder();
-            List<ShardRouting> shardRoutingsToBeApplied = new ArrayList<>(tasks.size());
-            for (ShardRoutingEntry task : tasks) {
-                shardRoutingsToBeApplied.add(task.shardRouting);
+        public BatchResult<ShardEntry> execute(ClusterState currentState, List<ShardEntry> tasks) throws Exception {
+            BatchResult.Builder<ShardEntry> builder = BatchResult.builder();
+            List<ShardEntry> validTasks = new ArrayList<>();
+            List<ShardRouting> validShardRoutings = new ArrayList<>(tasks.size());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: same naming comment and I liked toBeApplied better 👅

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 13]
```diff
@@ -370,27 +368,54 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
...
+            Set<ShardRouting> seenValidShardRoutings = new HashSet<>();
+            for (ShardEntry task : tasks) {
+                assert task.primaryTerm == 0L : "shard is only started by itself";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we add the task to the message?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 14]
```diff
@@ -370,27 +368,54 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
...
+
+                ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                if (matched == null) {
+                    // tasks that correspond to non-existent shards are marked as successful
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: this is a bit funky - can you add to the comment way it's ok (old stale requests that might be still in flight). My guess is thai is done partially because we resend shard started on every cluster state publishing that doesn't contain the shard as started so we have many of these in flight.   I think we will revisit this at some point - we shoudl have all the "channel failures" like master loss covered already.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 15]
```diff
@@ -370,27 +368,54 @@ public ShardStartedClusterStateTaskExecutor(AllocationService allocationService,
...
+                    logger.trace("{} shard with allocation id {} does not exist anymore, marking task {} as successful", task.shardId,
+                        task.allocationId, task);
+                    builder.success(task);
+                } else {
+                    if (matched.initializing() == false) {
+                        logger.trace("{} shard with allocation id {} exists, but is not initializing ({}), marking task {} as successful",
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we assert it's active?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 16]
```diff
@@ -429,111 +439,70 @@ private boolean deassociateDeadNodes(RoutingAllocation allocation) {
...
+
+            routingNodes.started(startedShard);
+            logger.trace("{} marked shard as started (routing: {})", startedShard.shardId(), startedShard);
+
+            if (startedShard.relocatingNodeId() != null) {
+                // relocation target has been started, remove relocation source
+                RoutingNode relocationSourceNode = routingNodes.node(startedShard.relocatingNodeId());
+                ShardRouting relocationSourceShard = relocationSourceNode.getByShardId(startedShard.shardId());
+                assert relocationSourceShard.isRelocationSourceOf(startedShard);
+                assert relocationSourceShard.getTargetRelocatingShard() == startedShard : "relocation target mismatch, expected: "
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: ah - now I see what you did it :)

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java [hunk 17]
```diff
@@ -429,111 +439,70 @@ private boolean deassociateDeadNodes(RoutingAllocation allocation) {
...
+            "shard routing to fail does not exist in routing table, expected: " + failedShard + " but was: " +
+                routingNodes.getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId());
+
+        logger.debug("{} failing shard {} with unassigned info ({})", failedShard.shardId(), failedShard, unassignedInfo.shortSummary());
         if (failedShard.primary()) {
             // fail replicas first otherwise we move RoutingNodes into an inconsistent state
             failReplicasForUnassignedPrimary(allocation, failedShard);
         }
 
-        if (addToIgnoreList) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: why is it safe to remove this?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 18]
```diff
@@ -214,71 +224,89 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                        continue;
+                    }
 
-            // tasks that correspond to non-existent shards are marked
-            // as successful
-            batchResultBuilder.successes(partition.getOrDefault(ValidationResult.SHARD_MISSING, Collections.emptyList()));
+                    ShardRouting matched = currentState.getRoutingTable().getByAllocationId(task.shardId, task.allocationId);
+                    if (matched == null) {
+                        // tasks that correspond to non-existent shards are marked as successful
+                        logger.trace("{} shard with allocation id {} does not exist anymore, marking task {} as successful", task.shardId,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: looking at the old code I see these are debug , and I tend to agree - how failures are handled is important (we long them on debug/warn on the other side). thoughts?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardFailedClusterStateTaskExecutorTests.java [hunk 19]
```diff
@@ -268,50 +269,11 @@ private static void assertTaskResults(
         }
     }
 
-    private static List<ShardStateAction.ShardRoutingEntry> toTasks(ClusterState currentState, List<ShardRouting> shards, String indexUUID, String message) {
+    private static List<ShardStateAction.ShardEntry> toTasks(ClusterState currentState, List<ShardRouting> shards, String indexUUID, String message) {
         return shards
             .stream()
-            .map(shard -> new ShardStateAction.ShardRoutingEntry(shard, randomValidSourceShard(currentState, shard), message, new CorruptIndexException("simulated", indexUUID)))
+            .map(shard -> new ShardStateAction.ShardEntry(shard.shardId(), shard.allocationId().getId(),
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardFailedClusterStateTaskExecutorTests.java]: shouldn't we keep the randomization between a local failure and a valid primary term?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java [hunk 20]
```diff
@@ -218,9 +220,6 @@ public void testFailPrimaryStartedCheckReplicaElected() {
         assertThat(routingTable.index("test").shard(0).primaryShard().currentNodeId(), anyOf(equalTo("node1"), equalTo("node2")));
         assertThat(routingTable.index("test").shard(0).replicaShards().size(), equalTo(1));
         assertThat(routingTable.index("test").shard(0).replicaShards().get(0).state(), equalTo(UNASSIGNED));
-
-        logger.info("fail the shard again, check that nothing happens");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java]: this made go and check that this is tested in ShardStateActionTests , which it is, but we miss test there, like double sending the same shard before it's processed and failing a shard of an index that doesn't exist

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/StartedShardsRoutingTests.java [hunk 21]
```diff
@@ -48,77 +47,35 @@ public void testStartedShardsMatching() {
...
-                .numberOfShards(3).numberOfReplicas(0)
+                .numberOfShards(2).numberOfReplicas(0)
                 .build();
         final Index index = indexMetaData.getIndex();
         ClusterState.Builder stateBuilder = ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY))
                 .nodes(DiscoveryNodes.builder().put(newNode("node1")).put(newNode("node2")))
                 .metaData(MetaData.builder().put(indexMetaData, false));
 
         final ShardRouting initShard = TestShardRouting.newShardRouting(new ShardId(index, 0), "node1", true, ShardRoutingState.INITIALIZING);
-        final ShardRouting startedShard = TestShardRouting.newShardRouting(new ShardId(index, 1), "node2", true, ShardRoutingState.STARTED);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/StartedShardsRoutingTests.java]: same here - we need move double starting and such to ShardStateActionTests

**Comment by bleskes** [comment]
I like a lot how this turned out. I left some minor comments but a few important ones.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 17 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: it was not removed. I moved it to the `applyFailedShards` method (https://github.com/elastic/elasticsearch/pull/19715/files#diff-792ac9952eb61524ecac9d045d6eae4bR248). That method was the only one calling `applyFailedShard` with `addToIgnoreList = true`.

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java (hunk 20 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java]: This is tested in ShardFailedClusterStateTaskExecutorTests. `testDuplicateFailuresAreOkay` checks duplicate failures (`createExistingShards` creates the duplicates using `int numberOfTasks = randomIntBetween(failures.size(), 2 * failures.size());`). `testNonExistentShardsAreMarkedAsSuccessful` fails a shard of non-existent index (`createNonExistentShards` uses existing shards with wrong allocation id, or wrong index etc.
In conclusion, the existing tests there should cover us.

**Comment by ywelsch** [comment]
@bleskes I've pushed another commit addressing review comments. It mainly contains improvements to the logging.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 22]
```diff
@@ -246,65 +246,65 @@ public ShardFailedClusterStateTaskExecutor(AllocationService allocationService,
...
-                            indexMetaData.primaryTerm(task.shardId.id()), task.primaryTerm, task);
-                        batchResultBuilder.failure(task, new NoLongerPrimaryShardException(
-                            task.shardId,
-                            "primary term [" + task.primaryTerm + "] did not match current primary term [" +
-                                indexMetaData.primaryTerm(task.shardId.id()) + "]"));
-                        continue;
+                    if (task.primaryTerm > 0) {
+                        long currentPrimaryTerm = indexMetaData.primaryTerm(task.shardId.id());
+                        if (currentPrimaryTerm != task.primaryTerm) {
+                            assert currentPrimaryTerm >= task.primaryTerm : "received a primary term with a higher term than in the " +
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: nit: `>=` -> `>` (equality is not an option :))

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java (hunk 17 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java]: ++

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java (hunk 20 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/FailedShardsRoutingTests.java]: Ok. great. Thanks for doing the research.

**Comment by bleskes** [comment]
LGTM. great work. 🎉

**Comment by ywelsch** [comment]
@bleskes Thanks for the review and the suggestion to resolve ShardRouting instances in ShardStateAction! This creates a clear separation of concerns between AllocationService and ShardStateAction.
